### PR TITLE
Add extra links for google dataproc

### PR DESCRIPTION
**Pull request by yesemsanthoshkumar**
Add extra links for google cloud operators - Dataproc

## How it is implemented
1. DataprocOperator pushes the details to xcom
2. The value pushed into Xcom is fetched to construct the link for job

**Note:** I haven't worked with dataproc workflows so I haven't implemented links for WorkFlow related operators.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by yesemsanthoshkumar** [comment]
@mik-laj 
I've added extra link for dataproc jobs.

However, I've the following questions.

1. While running the pre-commit hooks, I see both pylint and fake8 running checks. I've marked both of them to disable line too long error for some lines inside the files. Why do we have both of them?
2. I'm directing the link to the cluster if the job hasn't run and not generated any JobId. Is this approach valid? Do you want me to change this behaviour?
3. I coudn't find the cluster_name argument in [DataprocSubmitJobOperator](https://airflow.readthedocs.io/en/latest/_api/airflow/providers/google/cloud/operators/dataproc/index.html#airflow.providers.google.cloud.operators.dataproc.DataprocSubmitJobOperator) for which I've added the links. Shoud I add the links to [DataprocJobBaseOperator](https://airflow.readthedocs.io/en/latest/_api/airflow/providers/google/cloud/operators/dataproc/index.html#airflow.providers.google.cloud.operators.dataproc.DataprocJobBaseOperator)? Or Should I add on both?

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 1]
```diff
@@ -453,6 +477,50 @@ def test_execute(self, mock_hook):
...
+        op = DataprocSubmitJobOperator(
+            task_id=TASK_ID,
+            location=GCP_LOCATION,
+            project_id=GCP_PROJECT,
+            job=job,
+            gcp_conn_id=GCP_CONN_ID,
+            cluster_name=CLUSTER_NAME
+        )
+        self.dag.clear()
+        session.query(XCom).deete()
```
Comment (mik-laj) [inline tests/providers/google/cloud/operators/test_dataproc.py]: Deete=>delete

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 2]
```diff
@@ -39,13 +39,41 @@
...
+from airflow.models import BaseOperator, BaseOperatorLink
+from airflow.models.taskinstance import TaskInstance
 from airflow.providers.google.cloud.hooks.dataproc import DataprocHook, DataProcJobBuilder
 from airflow.providers.google.cloud.hooks.gcs import GCSHook
 from airflow.utils import timezone
 from airflow.utils.decorators import apply_defaults
 from airflow.version import version as airflow_version
 
+# pylint: disable=line-too-long
+DATAPROC_JOB_LOG_LINK = "https://console.cloud.google.com/dataproc/jobs/{job_id}?region={region}&project={project_id}"      # noqa: E501
```
Comment (michalslowikowski00) [inline airflow/providers/google/cloud/operators/dataproc.py]: Maybe this way?
```suggestion
DATAPROC_JOB_LOG_LINK = \
  "https://console.cloud.google.com/dataproc/jobs/{job_id}?region={region}&project={project_id}" 
```

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 3]
```diff
@@ -453,6 +477,50 @@ def test_execute(self, mock_hook):
...
+
+        ti = TaskInstance(
+            task=op,
+            execution_date=execution_date
+        )
+
+        self.assertEqual(
```
Comment (michalslowikowski00) [inline tests/providers/google/cloud/operators/test_dataproc.py]: Little bit over engineering but maybe this way:
Them `# pylint: disable=line-too-long` won't be needed anymore.
```

URL = "https://console.cloud.google.com/dataproc/"
cluster_name = CLUSTER_NAME
region = GCP_LOCATION
project_id = GCP_PROJECT


self.assertEqual(
  f'{URL}clusters/{cluster_name}/job?region={region}&project={project_id}'
  op.get_extra_links(execution_date, DataprocJobLink.name)
)
```

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 4]
```diff
@@ -453,6 +477,50 @@ def test_execute(self, mock_hook):
...
+                region=GCP_LOCATION,
+                project_id=GCP_PROJECT
+            ),
+            op.get_extra_links(execution_date, DataprocJobLink.name)
+        )
+
+        ti.xcom_push(key='job_id', value=job_id)
+
+        self.assertEqual(
+            # pylint: disable=line-too-long
```
Comment (michalslowikowski00) [inline tests/providers/google/cloud/operators/test_dataproc.py]: Same here with fstring.

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 5]
```diff
@@ -1704,6 +1743,8 @@ def execute(self, context: Dict):
         )
         self.log.info("Job completed successfully.")
 
+        context['task_instance'].xcom_push(key='job_id', value=job_id)
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py]:
```suggestion
        self.xcom_push(context, key='job_id', value=job_id)
```

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 6]
```diff
@@ -419,6 +428,24 @@ def test_execute(self, mock_hook):
...
+            'start_date': DEFAULT_DATE
+        })
+        self.mock_context = MagicMock()
+
+    def tearDown(self):
+        session = Session()
+        session.query(TaskInstance).filter_by(
+            dag_id=TEST_DAG_ID).delete()
+        session.commit()
+        session.close()
```
Comment (turbaszek) [inline tests/providers/google/cloud/operators/test_dataproc.py]: Let's use `setUpClass` so we do this logic once

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 7]
```diff
@@ -453,6 +480,70 @@ def test_execute(self, mock_hook):
...
+    def test_operator_extra_links(self, session):
+        job = {}
+        job_id = 'test_job_id_12345'
+        execution_date = datetime(2020, 7, 20)
+        # pylint: disable=line-too-long
+        expected_extra_link = 'https://console.cloud.google.com/dataproc/jobs/{job_id}?region={region}&project={project_id}'.format(     # noqa: E501
+            job_id=job_id,
+            region=GCP_LOCATION,
+            project_id=GCP_PROJECT
+        )
```
Comment (turbaszek) [inline tests/providers/google/cloud/operators/test_dataproc.py]:
```suggestion
        expected_extra_link = \
            f'https://console.cloud.google.com/dataproc/jobs/{job_id}?'
            f'region={GCP_LOCATION}&project={GCP_PROJECT}'
```
How about using f-string?

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 6 above)
Comment (turbaszek) [reply tests/providers/google/cloud/operators/test_dataproc.py]: And we can use `tests.tests_utils.db.clear_db_runs` (and `clear_db_xcom` as we push something to XCom) instead duplicating the logic

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 8]
```diff
@@ -590,7 +681,7 @@ def test_execute(self, mock_hook, mock_uuid):
             query=self.query,
             variables=self.variables,
         )
-        op.execute(context={})
+        op.execute(context=MagicMock())
```
Comment (turbaszek) [inline tests/providers/google/cloud/operators/test_dataproc.py:1136]: I suppose you did this because of adding `xcom_push` in `execute`, right?

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 9]
```diff
@@ -763,6 +854,23 @@ class TestDataProcSparkOperator(unittest.TestCase):
...
+            'start_date': DEFAULT_DATE
+        })
+        self.mock_context = MagicMock()
+
+    def tearDown(self):
+        session = Session()
+        session.query(TaskInstance).filter_by(
+            dag_id=TEST_DAG_ID).delete()
+        session.commit()
+        session.close()
```
Comment (turbaszek) [inline tests/providers/google/cloud/operators/test_dataproc.py]: Same here, we can use existing methods from `tests.test_utils.db`. And here we also should clear XCom to avoid side effect that is created on L953

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 2 above)
Comment (yesemsanthoshkumar) [reply airflow/providers/google/cloud/operators/dataproc.py]: Done

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 3 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Have moved the expected link to a global variable. Let me know if it needs further changes.

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 4 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Same as above

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 5 above)
Comment (yesemsanthoshkumar) [reply airflow/providers/google/cloud/operators/dataproc.py]: Done

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 6 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Have made those changes. One question though, I've added the tests in DataProcSubmiSparkOperator(DataprocJobBaseOperator). Should I add the tests for other subclass operators as well?

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 7 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Moved the expected link to a variable. Also converted it into f-string. Let me know if there is a better approach.

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 8 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py:1136]: Yes.

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 9 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Done using test_utils

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 6 above)
Comment (turbaszek) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Those operators are depecated so it's ok if we test with only one. But please add tests for `DataprocSubmitJobOperator`

**Review by turbaszek — COMMENTED** [review]
We should also add the console link to 
https://github.com/apache/airflow/blob/558be73ae808b551a8ccf59a5c1896b5718f2f44/airflow/serialization/serialized_objects.py#L44-L49

And as mentioned in a comment - we should add the link in `DataprocSubmitJobOperator` as other ops are deprecated. So, we may even consider to add this **only** to this one operator 😉

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 6 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Added already.
https://github.com/apache/airflow/pull/10343/files#diff-68c5fc800ea76a624d56445376a5744fR485

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 3 above)
Comment (michalslowikowski00) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Great. :)

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 4 above)
Comment (michalslowikowski00) [reply tests/providers/google/cloud/operators/test_dataproc.py]: Cool

**Comment by yesemsanthoshkumar** [comment]
> We should also add the console link to
> https://github.com/apache/airflow/blob/558be73ae808b551a8ccf59a5c1896b5718f2f44/airflow/serialization/serialized_objects.py#L44-L49
> 
> And as mentioned in a comment - we should add the link in `DataprocSubmitJobOperator` as other ops are deprecated. So, we may even consider to add this **only** to this one operator 😉

@turbaszek I've added to those other operator as well already.
:D

**Comment by yesemsanthoshkumar** [comment]
@mik-laj
> 2. I'm directing the link to the cluster if the job hasn't run and not generated any JobId. Is this approach valid? Do you want me to change this behaviour?

The link is not being redirected to the cluster URL if there is no detail of the job in the system.

> 3. I coudn't find the cluster_name argument in [DataprocSubmitJobOperator](https://airflow.readthedocs.io/en/latest/_api/airflow/providers/google/cloud/operators/dataproc/index.html#airflow.providers.google.cloud.operators.dataproc.DataprocSubmitJobOperator) for which I've added the links. Shoud I add the links to [DataprocJobBaseOperator](https://airflow.readthedocs.io/en/latest/_api/airflow/providers/google/cloud/operators/dataproc/index.html#airflow.providers.google.cloud.operators.dataproc.DataprocJobBaseOperator)? Or Should I add on both?

Implemented on both the operators. For subclasses of DataprocJobBaseOperators, I've added tests only for DataprocSubmitSparkJob operator as the other operators are anyway deprecated. Please have a look at the above comments by @turbaszek regarding this.

File: tests/providers/google/cloud/operators/test_dataproc.py (hunk 8 above)
Comment (yesemsanthoshkumar) [reply tests/providers/google/cloud/operators/test_dataproc.py:1136]: Have added a mock task instance for testing the xcom push. This change is regarding the operator extra links. Please take a look at the updated tests.

**Review by turbaszek — APPROVED** [review]
LGTM! 🚀

**Comment by turbaszek** [comment]
@yesemsanthoshkumar can you please rebase? We started using black on providers package 👍

**Comment by turbaszek** [comment]
Nice to see this as part of #10014

File: tests/providers/google/cloud/operators/test_dataproc.py [hunk 10]
```diff
@@ -434,7 +591,10 @@ def test_execute(self, mock_hook):
             request_id=REQUEST_ID,
             impersonation_chain=IMPERSONATION_CHAIN,
         )
-        op.execute(context={})
+        op.execute(context=self.mock_context)
+
+        # Test whether xcom push occurs before polling for job
+        self.extra_links_manager_mock.assert_has_calls(self.extra_links_expected_calls, any_order=False)
```
Comment (yesemsanthoshkumar) [inline tests/providers/google/cloud/operators/test_dataproc.py]: @turbaszek Have added checks for xcom_push before wait_for_job too. Please take a look.

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 11]
```diff
@@ -1051,17 +1051,17 @@ def execute(self, context):
                 project_id=self.project_id, job=self.job["job"], location=self.region,
             )
             job_id = job_object.reference.job_id
+            # XCom push is referenced by extra links and has to be before polling for job completion
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py]:
```suggestion
            # Save data required to display extra link no matter what job status will be
```
WDYT?

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 12]
```diff
@@ -627,6 +681,11 @@ def execute(self, context):
             cluster = self._create_cluster(hook)
             self._handle_error_state(hook, cluster)
 
+        self.xcom_push(
+            context,
+            key="cluster_conf",
+            value={"cluster_name": self.cluster_name, "region": self.region, "project_id": self.project_id,},
+        )
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py]: Let's move it to right after hook invocation. In this way, there will be cluster link no matter what is the state of the cluster. Only problem I see is when deleting cluster on error state but this is an edge case

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 11 above)
Comment (yesemsanthoshkumar) [reply airflow/providers/google/cloud/operators/dataproc.py]: Yeah. Clear than my comment. Will make changes.

File: airflow/providers/google/cloud/operators/dataproc.py (hunk 12 above)
Comment (yesemsanthoshkumar) [reply airflow/providers/google/cloud/operators/dataproc.py]: Similar flow with the jobs. Sure. Will update.

Comment (turbaszek) [reply airflow/providers/google/cloud/operators/dataproc.py]: Awesome, let me know once I can re-run DAG to check if the links work as expected

Comment (yesemsanthoshkumar) [reply airflow/providers/google/cloud/operators/dataproc.py]: @turbaszek Updated. Please take a look.

**Comment by turbaszek** [comment]
Tested both links. This rocks! Thanks @yesemsanthoshkumar 🚀

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 13]
```diff
@@ -41,13 +41,63 @@
...
+)
 from airflow.providers.google.cloud.hooks.gcs import GCSHook
 from airflow.utils import timezone
 from airflow.utils.decorators import apply_defaults
 from airflow.version import version as airflow_version
 
+DATAPROC_BASE_LINK = "https://console.cloud.google.com/dataproc"
+DATAPROC_JOB_LOG_LINK = DATAPROC_BASE_LINK + "/jobs/{job_id}?region={region}&project={project_id}"
+DATAPROC_CLUSTER_LINK = (
+    DATAPROC_BASE_LINK + "/clusters/{cluster_name}/monitoring?" "region={region}&project={project_id}"
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py]:
```suggestion
    DATAPROC_BASE_LINK + "/clusters/{cluster_name}/monitoring?region={region}&project={project_id}"
```

**Comment by yesemsanthoshkumar** [comment]
@turbaszek Should I rebase now before merging?

**Comment by turbaszek** [comment]
@yesemsanthoshkumar no need to rebase. However, we will need to wait with the merge until #10014 is done. That's because 1.10.2 and earlier versions will not support those new extra links. So this PR should be targeted to Airflow 2.0

**Review by ashb — DISMISSED** [review]
Please could you now rebase?

And additionally, please revert all the quote changes -- this introduces _a lot_ of changes unrelated to the PR subject which we like to avoid.

**Comment by turbaszek** [comment]
@ashb why do rebase now? We should wait with this till 2.0 otherwise we will break serialization one more time. Reference:
https://github.com/apache/airflow/issues/10171
https://github.com/apache/airflow/pull/10318

**Comment by ashb** [comment]
Oh sorry, I misread the conversation in this issue!

**Comment by potiuk** [comment]
@yesemsanthoshkumar  -> can you please rebase so that we could merge it before 2.0.0 rc1 ?

**Comment by potiuk** [comment]
REmove it from 2.0rc1 for now

**Comment by turbaszek** [comment]
@yesemsanthoshkumar could you take a look the CI issues? Some static checks seem to be failing.

**Comment by yesemsanthoshkumar** [comment]
@turbaszek Updated the PR. Please take a look and let me know of any further changes required.

@ashb I guess the quote changes(' -> ") you mentioned are the result of applying black to those files. Or are you referring to some other changes?

File: airflow/providers/google/cloud/operators/dataproc.py [hunk 14]
```diff
@@ -173,11 +218,11 @@ def __init__(
...
         num_masters: int = 1,
-        master_machine_type: str = 'n1-standard-4',
-        master_disk_type: str = 'pd-standard',
+        master_machine_type: str = "n1-standard-4",
+        master_disk_type: str = "pd-standard",
         master_disk_size: int = 1024,
-        worker_machine_type: str = 'n1-standard-4',
-        worker_disk_type: str = 'pd-standard',
+        worker_machine_type: str = "n1-standard-4",
+        worker_disk_type: str = "pd-standard",
```
Comment (turbaszek) [inline airflow/providers/google/cloud/operators/dataproc.py]: Unrelated changes, please revert all quote changes as asked previously by @ashb

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Wooho! :tada: -> took some time :)
